Sources/NIOHTTP1/NIOTypedHTTPServerUpgraderStateMachine.swift

```diff
@@ -382,5 +382,23 @@ struct NIOTypedHTTPServerUpgraderStateMachine<UpgradeResult> {
         }
     }
 
+    @usableFromInline
+    enum CloseInboundAction {
+        case close
+        case `continue`
+    }
+
+    @inlinable
+    mutating func closeInbound() -> CloseInboundAction {
+        switch self.state {
+        case .initial, .awaitingUpgrader:
```

Comment:
We should immediately close in the state `initial`. In the state `awaitingUpgrader` we have to check if we have `seenFirstRequest == true`. If have seen that we should buffer the `inputClosed` and send it after the upgrading is finished. In the state `upgraderReady` we need to buffer the `inputClosed` indefinitely.

The reason why this is so complicated is that the client could potentially send a half closure after it send the initial upgrade request + some data on the upgraded protocol. We can't just unconditionally close here but we rather need to buffer the `inputClose` and unbuffer it with the data. I would recommend changing the `var buffer: Deque<NIOAny>` in the various states to `var buffer: Deque<enum(NIOAny | inputClosed)>`.

Reply:
Done!